packages/xo-server/docs/guidelines.md

```diff
@@ -0,0 +1,284 @@
+- add an error code or a custom class extending error for custom errors
+
+```ts
+// ✅
+try {
+  db.insert();
+} catch (error) {
+  log.warn('DB full', { error });
+  throw;
+}
+try{
+
+}catch(err){
+   const specificError= new Error('this is a specific error')
+   specificError.cause = err
+   specificError.code = 'ENOENT'
+}
+
+
+// ❌
+JSON.parse(input); // No try/catch
+try{
+   asyncProcess() // no await => unbounded error
+}catch(error){}
+promise.catch(err=>throw err) // unbounded no stack trace
+
+// ⚠️
+try { parse(); } catch {} // Silenced error
+try{
+
+}catch(err){
+   throw new Error('this is a specific error')// lost the original error
+}
+```
+
+### Logging
+
+Use structured logs with `@xen-orchestra/log` avoid flooding.
+
+```ts
+// ✅
+log.info('VM started', { vmId, durationMs })
+
+// ❌
+console.log(`${vm} started at ${new Date()}`) // Unstructured
+
+// ⚠️
+log.error(err) // No context like VM ID
+ayncPromise.catch(() => {}) // erro is forcibly silenced
+```
+
+### Readability
+
+- name method and variable explicitly, especially on non typed code. For example, set/edit method should only be used on a typed object. Avoid generic code like for (const object of objects){}
+- method parameter are readonly and explicit. They are typed.
+- function name are readonly and explicit.
+- limit number of parameters or use oject parameter
+
+```ts
+// ✅
+for (const vm of runningVms) {
+  await vm.stop()
+}
+function work(ref, { bootable, seekable, focusable = true, otherboolean = false } = {}) // work(1, {bootable: true , focusable:false})
+// ❌
+for (const x of y) {
+  /* What's x? */
+}
+
+let set = (item, key, value) => (item[key] = value)
+set = item => set(item, 'mykey', 4)
+
+// ⚠️
+items.forEach(item => process(item)) // 'item' too vague
+function work(ref, bootable, seekable, focusable, otherboolean) //work(1, true, true, false,true) or work(1, true, false, false,true)
+```
+
+### Scalability
+
+No unbounded memory/parallel ops.
+
+```ts
+// ✅
+for await (const chunk of stream) {
+  process(chunk)
+}
+
+// ❌
+const allData = await fs.readFile('50TB.log')
+
+// ⚠️
+await Promise.all(users.map(fetchDetails)) // 50K parallel requests
+```
+
+### Testing
+
+Use `node:assert`; test critical paths.
+
+```js
+// ✅
+assert.strictEqual(result.status, 'running')
+
+// ❌
+if (result !== 'running') throw Error() // Ad-hoc
+
+// ⚠️
+assert(result) // No error message
+```
+
+### Typing
+
+Zero `any`; JSDoc/TypeScript everywhere.
+
+```ts
+// ✅
+/** @type {Array<{id: string}>} */
+const vms = []
+
+// ❌
+const data = {} // Untyped
+
+// ⚠️
+function parse(input: any) {} // Leaky abstraction
+```
+
+Node 23 is around the corner and support a subset of typescript. It doesn't support enums, namespaces and parameters properties. Also is need to use the `type` keywork when importing type. If a module don't need this feature, it will be able to skip the compilation phase at build and keep the type checking in developpment mode
```

Comment:
```suggestion
Node 23 supports a subset of typescript. It doesn't support enums, namespaces and parameters properties. Also is need to use the `type` keywork when importing type. If a module don't need this feature, it will be able to skip the compilation phase at build and keep the type checking in developpment mode
```